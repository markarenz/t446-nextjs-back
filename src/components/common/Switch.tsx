import styles from '../../styles/modules/Switch.module.scss';

type Props = {
  name: string;
  tabIndex: number;
  value: string;
  onChange: Function;
  options: {
    label: string;
    value: string;
  }[];
};

const Switch: React.FC<Props> = ({
  name,
  tabIndex,
  value,
  onChange,
  options
}) => {
  const handleClick = () => {
    const newVal = value === 'inactive' ? 'active' : 'inactive';
    onChange(name, newVal);
  };
  if (options.length < 2) {
    return null;
  }
  const [optionA, optionB] = options;
  const isOptionB = value === optionB.value;
  return (
    <div className={styles.switchRoot}>
      <span className={styles.optionLabel}>{optionA.label}</span>
      <button
        type="button"
        onClick={handleClick}
        tabIndex={tabIndex}
        className={styles.switchBtn}
        title="switch"
        aria-label="switch"
      >
        <svg viewBox="0 0 100 100" className={isOptionB ? styles.optionB : ''}>
          <circle cx="50" cy="50" r="50" className={styles.circleOuter} />
          <circle cx="50" cy="50" r="40" className={styles.circleInner} />
        </svg>
      </button>
      <span className={styles.optionLabel}>{optionB.label}</span>
    </div>
  );
};

export default Switch;
