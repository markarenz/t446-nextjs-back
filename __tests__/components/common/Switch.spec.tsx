import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Switch from '../../../src/components/common/Switch';

const defaultProps = {
  name: 'fieldName',
  tabIndex: 1,
  value: 'active',
  onChange: jest.fn(),
  options: [
    { label: 'On', value: 'active' },
    { label: 'Off', value: 'inactive' }
  ],
  testId: 'test-switch'
};
const defaultProps2 = {
  ...defaultProps,
  value: 'inactive'
};
const altProps = {
  ...defaultProps,
  value: 'inactive',
  options: [{ label: 'Off', value: 'inactive' }]
};
describe('Switch', () => {
  it('renders Switch', async () => {
    render(<Switch {...defaultProps} />);
    const element = await screen.findByTestId('test-switch');
    expect(element).toBeInTheDocument();
  });
  it('renders nothing when fewer than 2 options are provded', async () => {
    render(<Switch {...altProps} />);
    const element = screen.queryByTestId('test-switch');
    expect(element).toBe(null);
  });
  it('fires onChange when clicked - active', async () => {
    render(<Switch {...defaultProps} />);
    const button = await screen.findByTestId('test-switch-button');
    fireEvent(
      button,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );
    expect(defaultProps.onChange).toHaveBeenCalled();
  });
  it('fires onChange when clicked - inactive', async () => {
    render(<Switch {...defaultProps2} />);
    const button = await screen.findByTestId('test-switch-button');
    fireEvent(
      button,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true
      })
    );
    expect(defaultProps.onChange).toHaveBeenCalled();
  });
});
