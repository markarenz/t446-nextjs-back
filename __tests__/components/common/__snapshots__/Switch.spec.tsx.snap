// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders Switch 1`] = `
<div
  class="switchRoot"
  data-testid="test-switch"
>
  <span
    class="optionLabel"
  >
    On
  </span>
  <button
    aria-label="switch"
    class="switchBtn"
    data-testid="test-switch-button"
    tabindex="1"
    title="switch"
    type="button"
  >
    <svg
      class=""
      viewBox="0 0 100 100"
    >
      <circle
        class="circleOuter"
        cx="50"
        cy="50"
        r="50"
      />
      <circle
        class="circleInner"
        cx="50"
        cy="50"
        r="40"
      />
    </svg>
  </button>
  <span
    class="optionLabel"
  >
    Off
  </span>
</div>
`;
